### PR TITLE
feat(rocm): declarative per-op/per-arch capability table

### PR DESCRIPTION
**Pull request by demandal25**
## Summary

Architecture gating today is a single boolean allowlist — `FLASHINFER_SUPPORTED_ROCM_ARCHS` — so no code path distinguishes gfx942 from gfx950, and that same list is reused by `is_aiter_supported()`. Any architecture added for in-tree HIP kernels silently becomes "AITER-supported" too. This adds the table that makes support explicit per `(op, backend, arch)`.

**No behaviour change.** Nothing consumes the table yet; routing the three divergent AITER validators through it is a follow-up.

### What changed

- **`flashinfer/arch_caps.py`** — `Support`, `KnownBad`, `ArchSupport`, `Capability`, the `CAPABILITIES` table (24 rows), `ArchCapabilityError`, and the query API (`capability_available` / `require_capability`).
- **`tests/rocm_tests/test_arch_caps_hip.py`** — 53 tests across table well-formedness, immutability, version-window arithmetic, gating, version-probe cost, and the exception contract.

### Architecture / design notes

**Support is not a per-arch enum, because the defect we have is not architectural.** The one measured CDNA4 problem is a *compiler* bug: same architecture, same amd-aiter, correct on ROCm 7.1 and wrong on 7.2.x. A per-arch status can only record one of those. So a row carries a binary gate plus optional `KnownBad` version windows, with literal bounds so the window is inspectable, renderable into docs, and testable without a GPU.

Measured on an MI350X, torch and amd-aiter held constant, ROCm the only variable:

```
ROCm 7.2.0   max_abs_err vs fp32 reference 0.595268   3/12 tests fail
ROCm 7.2.4   max_abs_err vs fp32 reference 0.595268   bit-identical, four patches later
ROCm 7.1     max_abs_err vs fp32 reference 0.000250   12/12 tests pass
```

Upstream calls this a compiler bug and gates their own test skips on exactly `(major, minor) == (7, 2)` with gfx950 (`op_tests/test_batch_prefill.py::should_skip_rocm72_issue`), still present at aiter **v0.1.20** — so no amd-aiter upgrade avoids it. The window here was confirmed by measurement rather than taken from that comment.

**Rows carry `evidence`, not a "validated" flag.** Each entry names what was actually run — board / ROCm / AITER / torch / date. The AITER rows are populated from full-suite runs on both architectures at `49a8cdd8`: gfx942 **27520 passed / 0 failed**, gfx950 **27517 passed / 3 failed** (those 3 being the defect above). HIP rows are declared but their evidence is deliberately **left empty** — the suites cover them, but no per-op HIP attribution was recorded, and borrowing the AITER runs' credibility would be exactly the unearned claim this table exists to remove.

**An architecture absent from a row resolves to unsupported**, so adding one to the allowlist grants nothing until it is declared. A conformance test enforces the converse — every allowlisted architecture must appear in every row — so adding e.g. gfx1200 fails CI until someone states what it can do. That closes the "any new arch is AITER-supported for free" hole.

**`ArchCapabilityError` subclasses `ValueError` and `RuntimeError`** so the three divergent exception types can collapse to one without breaking either existing catcher (`test_activation_aiter_hip.py:67` asserts `ValueError`; `test_batch_prefill_bf16_custom_mask_hip.py:157` catches `RuntimeError`).

It is deliberately **not** a `utils.GPUArchitectureError`, though that is the tidier hierarchy. That class lives in `flashinfer/utils.py`, which imports torch at module scope, and this module must stay torch-free — `hip_utils` imports it before `tests/conftest.py` pins `HIP_VISIBLE_DEVICES`. Re-declaring the base locally was tried and rejected: a same-named copy is a different class, so `except GPUArchitectureError` would not catch it, which is worse than not claiming the relationship at all. No ROCm path raised that type before, so nothing regresses and `skip_on_gpu_arch_error` remains CUDA-only.

**The table is immutable, and `device` leads the query signature.** `frozen=True` on `Capability` only stops the fields being rebound — a plain dict in `archs` would still let any importer edit the global table in place, so `__post_init__` coerces it to a `MappingProxyType`. The coercion sits on the class rather than on the row-construction helper, so it holds however a row is built. `ArchSupport` and `KnownBad` are already frozen with tuple fields, so nothing below that point is mutable either. Separately, `capability_available` / `require_capability` take `device` first to share a prefix with `aiter_utils.require_aiter(device, op)`: `op` and `backend` are both `str`, so a slip between them would be silent, and this keeps every migration to "append `backend`".

**Version detection is probed at most once.** `get_system_rocm_version()` walks up to four methods, three of which shell out (`amd-smi`, `dpkg`, `hipconfig`) with timeouts — too expensive to repeat per routing decision. It is now `lru_cache`d, and `_blocking_reason` skips it entirely for rows with no `known_bad` window, which is 23 of the 24 rows. Only the one gated row pays anything at all.

**The escape hatch is `FLASHINFER_ARCH_ALLOW_KNOWN_BAD=1`** — opt in to danger, never opt in to safety. An unreadable version does not match a window, so a machine whose ROCm we cannot detect is not gated on a guess.

### CDNA3 impact

None. Nothing calls into the table yet, and gfx942 is `SUPPORTED` for every declared row with no `known_bad` window.

## Test plan

Verified on gfx950 (MI350X, ROCm 7.2.0, torch 2.9.1+rocm7.2.0, amd-aiter 0.1.10):

- [x] 53 new tests pass.
- [x] `test_hip_utils.py` passes unchanged (69 tests).
- [x] The gating tests monkeypatch the architecture, so **gfx942 behaviour is exercised on a machine that has no gfx942** — including that gfx942 stays open on ROCm 7.2 where gfx950 is gated, which is the whole reason the window keys on both.
- [x] The module remains torch-free, enforced by a subprocess test that loads it by file path (importing via the package would run `flashinfer/__init__.py`, which pulls in torch, and prove nothing).
- [x] `pre-commit run -a`

### CODE REVIEW

File: flashinfer/arch_caps.py [hunk 1]
```diff
@@ -45,3 +59,313 @@ def normalize_arch(gcn_arch_name: str) -> str:
...
+        "ROCm 7.2.x miscompiles AITER's causal batch-prefill kernel on gfx950: "
+        "causal=True with logits_soft_cap=0.0 returns wrong numbers (not an "
+        "error), 97.6% of elements off. Use ROCm 7.1, or backend='fa2'"
+    ),
+    url="https://github.com/ROCm/aiter/blob/main/op_tests/test_batch_prefill.py",
+)
+
+
+def _aiter(gfx942: ArchSupport, gfx950: ArchSupport) -> Mapping[str, ArchSupport]:
+    return {"gfx942": gfx942, "gfx950": gfx950}
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/arch_caps.py]: Fixed in 399aefc6. Put the coercion on `Capability.__post_init__` rather than on the row helper, so a directly-constructed `Capability("x", "y", {...})` is protected too; `ArchSupport`/`KnownBad` are already frozen with tuple fields, so the whole structure is now read-only. Two tests added.

File: flashinfer/arch_caps.py [hunk 2]
```diff
@@ -45,3 +59,313 @@ def normalize_arch(gcn_arch_name: str) -> str:
...
+            rocm = get_system_rocm_version()
+    except Exception:
+        pass
+    try:
+        import importlib.metadata as _md
+
+        aiter = _md.version("amd-aiter")
+    except Exception:
+        pass
+    return rocm, aiter
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/arch_caps.py]: Fixed in 399aefc6. `_live_versions` is now `lru_cache`d, and `_blocking_reason` skips it entirely when the row has no `known_bad` window — 23 of the 24 rows, so they cost nothing rather than being merely cheap after the first call. Two tests added.

File: flashinfer/arch_caps.py [hunk 3]
```diff
@@ -45,3 +59,313 @@ def normalize_arch(gcn_arch_name: str) -> str:
...
+
+def require_capability(op: str, backend: str, device) -> None:
+    """Raise :class:`ArchCapabilityError` if ``backend`` cannot serve ``op`` here.
+
+    Mirrors ``aiter_utils.require_aiter`` -- same ``(device, op)`` information,
+    one exception type instead of three.
+    """
+    reason = _blocking_reason(op, backend, _device_arch(device))
+    if reason is not None:
+        raise ArchCapabilityError(reason)
```
[the remark this answers is not in the document]
Comment (demandal25) [reply flashinfer/arch_caps.py]: Fixed in 399aefc6 — both are now `(device, op, backend)`. Sharper than stated: `op` and `backend` are both `str`, so a slip between them is silent, and device-first keeps every migration to appending `backend` (`require_aiter(dev, "rmsnorm")` → `require_capability(dev, "rmsnorm", "aiter")`). No other call sites exist yet; only `normalize_arch` is imported elsewhere.

**Comment by demandal25** [comment]
**Suppressed comments in review 4978661573 — both accepted, fixed in ecefc2be.**

- **`test_arch_caps_hip.py:96` — hard-coded `SUPPORTED_ARCHS`.** Correct, and the drift is asymmetric in a way worth naming: a new arch would have been caught by `test_every_supported_arch_declared_in_every_row` but silently skipped by the routing parametrization, so the allowlist would have gained an arch that nothing actually exercised. Now derived from `FLASHINFER_SUPPORTED_ROCM_ARCHS`; `hip_utils` is torch-free at module scope, so the import hoists and the two function-local copies go away.

- **`arch_caps.py:297` — `_BY_KEY` drops duplicates.** Accepted. `test_keys_are_unique` did cover it, but a table whose entire purpose is to stop unearned support claims should not resolve a contradiction by silently picking one row. `_index()` now raises, with a test.

**Comment by demandal25** [comment]
**Suppressed comments in review 4978733669 — both accepted, fixed in 7e7d896e.**

- **`arch_caps.py:39` — `Capability` missing from `__all__`.** Correct; `CAPABILITIES` is exported and every element of it is one. Added, plus a test that `__all__` covers every public name defined in the module, so the next omission fails instead of lurking.

- **`arch_caps.py:116` — `"7.2" < "7.2.0"`.** Checked before accepting, and it is reachable rather than hypothetical: `get_system_rocm_version_from_hipconfig` matches `\d+\.\d+(?:\.\d+)?`, so the patch component is optional, and `get_system_rocm_version` consults *only* that method on TheRock builds (`hip_utils.py:167`) — a bare `"7.2"` is a value we can actually be handed. Comparison now zero-pads to equal width. To be clear about scope: the current table writes `rocm_min="7.2"` and gates correctly either way, so this is hardening for the next window, not a live-bug fix. Six parametrized cases added covering padding equality and both bounds.
